### ISSUE TITLE
The "Dynamism limitations" and "Inheritance Limitations" sections in Styling Local Dom are hard to follow

### ISSUE THREAD

**Issue by mercmobily** [Polymer/docs#1533]
# Dynamism limitations

I don't like writing issues in the documentation without fixing them (you should know that by now). However, for this one, all I can provide is code that shows that what I *think* the section is trying to say is wrong.

Here is the code:

````
<link rel="import" href="/bower_components/polymer/polymer.html">

<dom-module id="my-toolbar">
  <template>
    <style>
      :host {
        padding: 4px;
        background-color: gray;
      }
      .title {
        color: var(--my-toolbar-title-color);
      }
    </style>
    <span class="title">{{title}}</span>
  </template>
  <script>
    Polymer({
      is: 'my-toolbar',
      properties: {
        title: String
      }
    });
  </script>
</dom-module>

<dom-module id="x-custom">
  <template>
    <style>
      :host {
        --my-toolbar-title-color: black;
      }
      .funky {
        --my-toolbar-title-color: red;
      }
    </style>
    <my-toolbar id="normal" title="My awesome app, normal"></my-toolbar>
    <my-toolbar id="mega" class="funky" title="My awesome app, super"></my-toolbar>

    <button on-tap="changeTheme">Change theme</button>
  </template>
  <script>
    Polymer({
      is: 'x-custom',
      changeTheme: function() {
        this.$.normal.className += " funky"
        this.updateStyles();
      }
    });
  </script>
</dom-module>
````
So, adding a class doesn't change the color on the page (try to run this code without `updateStyles()`...)

Or maybe I totally misunderstood this section which had the following problems:

* No code to run (!)
* It's unclear what the limitation really is
* It seems to contradict what's been said before

# Inheritance limitations

I don't understand the limitations... and to be honest this part simply confused me.
This sentence: 

> "Unlike normal CSS inheritance which flows from parent to child, custom properties in Polymer’s shim can only change when inherited by a custom element from rules that set properties in scope(s) above it, or in a :host rule for that scope"

I am sure it's formally correct, and once I understand things I might be able to follow it, but while this section actually has code, it's a totally different type of element which in the local DOM's style part it sets `--custom-color` and then a CSS rule within there also _uses_ `--custom-color` as an attribute (?!?). 

This part:

> Unlike normal CSS inheritance which flows from parent to child, custom properties in Polymer’s shim can only change when inherited by a custom element from rules that set properties in scope(s) above it, or in a :host rule for that scope. 

Tells me that custom properties need to be changed by rules in the scope *above* it, or ::host. But so, why would this work?!? (and it does)

````
<link rel="import" href="/bower_components/polymer/polymer.html">

<dom-module id="my-toolbar">
  <template>
    <style>
      :host {
        padding: 4px;
        background-color: gray;
      }
      .title {
        color: var(--my-toolbar-title-color);
      }
    </style>
    <span class="title">{{title}}</span>
  </template>
  <script>
    Polymer({
      is: 'my-toolbar',
      properties: {
        title: String
      }
    });
  </script>
</dom-module>


<dom-module id="x-custom">
  <template>
    <style>
      :host {
        --my-toolbar-title-color: black;
      }
      .funky {
        --my-toolbar-title-color: red;
      }
    </style>
    <my-toolbar id="normal" title="My awesome app, normal"></my-toolbar>
    <my-toolbar id="mega" class="funky" title="My awesome app, super"></my-toolbar>

    <button on-tap="changeTheme">Change theme</button>
  </template>
  <script>
    Polymer({
      is: 'x-custom',
      changeTheme: function() {
        this.$.normal.className += " funky"
        this.updateStyles();
      }
    });
  </script>
</dom-module>
````
Aren't I setting `--my-toolbar-title-color` in two different spots, once of which is _not_ `:host`?

...? And then this sentence, in **bold**!

> Within a given element’s local DOM scope, a custom property can only have a single value.

Again... che?!? :D
Here, I would use an example so that the code is similar to the examples earlier in the chapter, highlighting what works and what *doesn't* more clearly. Again, sorry I can't provide a better writeup -- at this stage, I am not sure what the Dynamism and the Inheritance limitations are... 

Sorry, I don't mean to be a pain, but while I realise I am not really a bright reader, I don't think these two sections are as clear as they could be.

**Comment by arthurevans** [comment]
These sections need a rewrite, yes. 

As far as dynamism goes, the point is that _defining a property is not dynamic_:

      :host {
        --my-toolbar-title-color: black;
      }
      .funky {
        --my-toolbar-title-color: red;
      }

_Applying a property_ is dynamic:

      :host {
        color: var(--my-toolbar-title-color);
      }
      :host(.funky) {
        color: var(--my-toolbar-funky-title-color);
      }

http://jsbin.com/hoyiyoj/1/edit?html,output

The main point with inheritance is:

Inside one element's shadow DOM, the property can only have one value.

The trick here is that you're setting the property directly onto a custom element (my-toolbar), so in _that scope_, the property can take on a new value.

On the other hand, if you replace the toolbars with plain old divs, you see the issue:

http://jsbin.com/nohopi/1/edit?html,output

These two divs are _always_ going to be the same color.

**Comment by mercmobily** [comment]
Questions:

1)
Does my first example break the rule "Inside one element's shadow DOM, the property can only have one value" when I type:

      :host {
        --my-toolbar-title-color: black;
      }
      .funky {
        --my-toolbar-title-color: red;
      }

2)
When you say "Applying a property is dynamic:", can I ask "why?". BEFORE you answer, I will try and answer myself. In my example, I wrote:

<link rel="import" href="/bower_components/polymer/polymer.html">

<dom-module id="my-toolbar">
  <template>
    <style>
      :host {
        padding: 4px;
        background-color: gray;
      }
      .title {
        color: var(--my-toolbar-title-color);
      }
    </style>
    <span class="title">{{title}}</span>
  </template>
  <script>
    Polymer({
      is: 'my-toolbar',
      properties: {
        title: String
      }
    });
  </script>
</dom-module>

<dom-module id="x-custom">
  <template>
    <style>
      :host {
        --my-toolbar-title-color: black;
      }
      .funky {
        --my-toolbar-title-color: red;
      }
    </style>
    <my-toolbar id="normal" title="My awesome app, normal"></my-toolbar>
    <my-toolbar id="mega" class="funky" title="My awesome app, super"></my-toolbar>

    <button on-tap="changeTheme">Change theme</button>
  </template>
  <script>
    Polymer({
      is: 'x-custom',
      changeTheme: function() {
        this.$.normal.className += " funky"
        this.updateStyles();
      }
    });
  </script>
</dom-module>

When `<x-custom>` is created, it sets the value to assign to `--my-toolbar-title-color` for the host as a whole (`black`), and for anything with class `.funky` (`red`). When I instance `<my-toolbar>`, each element is statically assigned a value for `--my-toolbar-title-color`.

(Q: Does polymer do this "by hand" so to speak? It actually checks if each instanced element, in this case `<my-toolbar>`, fits in a selector and assign --my-toolbar-color accordingly...? And once again, does this code somehow "break" the rule you mentioned earlier, "one value per property inside a shadow DOM?).

Anyway, if one of the `<my-toolbar>` elements is assigned a different class later on (with a tap, a JS event, etc.), that's "too late" -- the widget has already been instanced with the "wrong" `--my-toolbar-title-color` and you need to run `this.updateStyles()` which (I assume) will check the selector of each sub-element of an element, and will change `--my-toolbar-title-color` (right?).
 
Doing it "your" way, however (I won't make this answer unnecessarily long pasting your JSBIN), you create *two* different values; so, when each `<my-toolbar>` is created, it has two neat values, and f their classes change, well, the change is picked up by the browser.

Please tell me I got it right :D Or, please correct the writeup of what Polymer does with something that is actually true...

3) 
In your SBIN, shouldn't `:host(.funky) .title {` be `:host(.funky) ::content .title {` to be "formally" correct and compatible with Shadow DOM? Not trying to criticise you :D Just checking that I got that part right too.

I am not going to torture you with more questions about inheritance, since this message is already long enough, AND I think the issues are interconnected.

--------------------
**Suggestion on this section**

In terms of structuring the chapter for a rewrite, I would follow an outline that would include:

* The "internals" of how Polymer deals with custom properties. Not in a very detailed way, but an explanation that sheds light on the "whys"

* A bunch of best practices. Since this is very much pattern-based, I would really, really have a bunch of "this is the best way to go about this" examples -- as well as "don't"s that only work in specific situations

I really wish we could have these in the specs -- we wouldn't have to deal with quite so many explanations, edge cases, Polymer-specific limitations, etc. if custom properties were part of most browsers (although it's a very fluid spec, with `/deep/` and `::shadow` deprecated so quickly... so yes.

Sorry about the obscenely long comments, BTW. I am just trying to "get it"... if the chapter is clear enough so that *I* get it, I guarantee *everybody* will be able to understand it too!

**Comment by mercmobily** [comment]
Arthur, did I get it all wrong in my comment above? :D

**Comment by mercmobily** [comment]
Eureka! Ignore question (1). I actually had a good hour to flay with this and fully understand it. As usual, once I know it, the docs now make a lot of sense. I would love to try and write some of this, although it's pretty complex (I might end up writing something way too long for anybody's good)

**Comment by mercmobily** [comment]
Please also ignore (3) -- of course not, you a styling the local DOM, not the light one. Sorry. I can't believe I still get confused between local and light DOM.

**Comment by mercmobily** [comment]
I guess this can be closed in favour of #1537 ... Which attempts to solve _all_ of these issues, and more. I hope I got it right... closing this now.
